### PR TITLE
Implement delivering client assets for custom elements in shared questions

### PR DESCRIPTION
**Pull request by SethPoulsen**
partial replacement for #9776 

resolves part of #9397

### CODE REVIEW

**Comment by SethPoulsen** [comment]
Do we have any automated tests for the `elementFiles` endpoint that I should be updating for this?

File: apps/prairielearn/src/question-servers/freeform.js [hunk 1]
```diff
@@ -1538,10 +1538,22 @@ export async function render(
         scriptUrls.push(assets.coreElementAssetPath(file)),
       );
       dependencies.courseElementStyles.forEach((file) =>
-        styleUrls.push(assets.courseElementAssetPath(course.commit_hash, locals.urlPrefix, file)),
+        styleUrls.push(
+          assets.courseElementAssetPath(
+            course.commit_hash,
+            locals.urlPrefix + question.course_id !== course.id ? `/variant/${variant.id}` : '',
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/question-servers/freeform.js:1544]: I believe you want brackets here (addition has higher precedence than ternary conditional):
```suggestion
            locals.urlPrefix + (question.course_id !== course.id ? `/variant/${variant.id}` : ''),
```

File: apps/prairielearn/src/pages/elementFiles/elementFiles.js [hunk 2]
```diff
@@ -53,8 +57,21 @@ router.get(
...
       // Files should be served from the course directory
-      const coursePath = chunks.getRuntimeDirectoryForCourse(res.locals.course);
-      await chunks.ensureChunksForCourseAsync(res.locals.course.id, { type: 'elements' });
+      let question_course;
+      if (req.params.variant_id) {
+        const variant = await selectVariantById(req.params.variant_id);
+        if (!variant || !idsEqual(variant.course_id, res.locals.course.id)) {
+          // the existence of the variant within the course validates that this course has sharing permissions on this question
+          throw new HttpStatusError(404, 'Not Found');
+        }
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/elementFiles/elementFiles.js:66]: There is no validation here that the user indeed has access to the variant in the first place. If a user is able to get ahold of (or guess) a variant ID that happens to be in the course they want, they would get access to the course element files. Also, I don't know how much validation would be desirable, but this would give someone access to all of the course's elements, not just the elements of the question being shared.

File: apps/prairielearn/src/question-servers/freeform.js (hunk 1 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/question-servers/freeform.js:1544]: That said: is there any advantage with using path parameters vs query parameters here? In other words, why not use `/pl/course/XXX/cacheableElements/YYY?variant_id=ZZZ`? That would reduce the need for changes in server.js. Unless you want to do some authorization check based on variants and prefixes, which is currently not done.

Comment (jonatanschroeder) [reply apps/prairielearn/src/question-servers/freeform.js:1544]: Another issue: this is done for any question, not just shared questions, which means the caching of elements in local browsers becomes linked to the variant, and you lose a lot of potential for caching.

**Review by jonatanschroeder — COMMENTED** [review]
I don't think there is an easy solution to the questions I'm bringing here, we can discuss it in our meeting.

File: apps/prairielearn/src/question-servers/freeform.js (hunk 1 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/question-servers/freeform.js:1544]: ah shoot, I totally missed the cacheing thing. Ya we should probably do this differently.

My main concern, aren't query params usually for smaller changes to what an endpoint means? Having the files come from a different course seems like a big change that should be a path param. But I guess I don't really know what our conventions are around path vs. query params.

File: apps/prairielearn/src/pages/elementFiles/elementFiles.js (hunk 2 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:66]: > no validation here that the user indeed has access to the variant in the first place

ok I'm embarrassed that I missed this but you are right. I'm not sure what the best way to do this is. The new way seems to be doing auth within the endpoints, where the old way was using middleware. I could just add another if statement here, not sure if there is a better solution.

> this would give someone access to all of the course's elements, not just the elements of the question being shared

yes, I was aware of this one. I'm not sure if this is a problem though. A few alternatives would be: 
- have the sharing course explicitly declare somewhere which elements they are willing to share
- parse the generated variant to see which elements it is using
both of these options seem like a total pain, and not worth the trouble.

Comment (mwest1066) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:66]: Ideally we'd like to reuse authz functions from the middleware. I'd like to avoid hand-crafted authz checks if at all possible. We basically want to perform the usual checks for access to the question associated with the variant? And check that it's shared?

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:66]: Also, only use the variant in the URL if the question course does not match the variant course, so caching can be used and this check is not required.

Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:66]: I'm not really sure what you mean by "reuse authz functions from the middleware", there aren't reusable functions, almost everything is done directly in the handlers. The closest thing to what you are saying is this: 
https://github.com/PrairieLearn/PrairieLearn/blob/04ccc99345977bfda77c20dc21e53665545d1106/apps/prairielearn/src/middlewares/selectAndAuthzInstanceQuestion.js#L11-L20

I can give it a try but it will be a little awkward because I will need to attach some things to locals for it to work, and it will throw as soon as it hits any issues.

File: apps/prairielearn/src/question-servers/freeform.js (hunk 1 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/question-servers/freeform.js:1544]: All fixed.

File: apps/prairielearn/src/question-servers/freeform.js [hunk 3]
```diff
@@ -1538,10 +1538,16 @@ export async function render(
         scriptUrls.push(assets.coreElementAssetPath(file)),
       );
       dependencies.courseElementStyles.forEach((file) =>
-        styleUrls.push(assets.courseElementAssetPath(course.commit_hash, locals.urlPrefix, file)),
+        styleUrls.push(
+          assets.courseElementAssetPath(course.commit_hash, locals.urlPrefix, file) +
+            (question.course_id !== course.id ? `?variant_id=${variant.id}` : ''),
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/question-servers/freeform.js:1543]:
```suggestion
            (!idsEqual(question.course_id, course.id) ? `?variant_id=${variant.id}` : ''),
```
same below.

File: docs/questionSharing.md [hunk 4]
```diff
@@ -47,7 +47,7 @@ If a sharing course attempts to share a question which accesses client or server
...
 
-Just as anyone with access to a question in your course can access any the `clientFilesQuestion`, anyone with permissions to any of the questions you have shared from your course may also access these `clientFilesQuestion`. These means that if you have any questions from your course that are publicly shared, anyone with access to the internet can access the `clientFilesQuestion` directories for these questions.
+Just as anyone with access to a question in your course can access any the `clientFilesQuestion`, anyone with permissions to any of the questions you have shared from your course may also access these `clientFilesQuestion`. These means that if you have any questions from your course that are publicly shared, anyone with access to the internet can access the `clientFilesQuestion` directories for these questions. Additionally, anyone with access to a question shared from your course will be able to access the client assets in your custom course elements, just as any students with access to your course can access the client assets in your custom course elements.
```
Comment (jonatanschroeder) [inline docs/questionSharing.md:50]:
```suggestion
Just as anyone with access to a question in your course can access any file in `clientFilesQuestion`, anyone with permissions to any of the questions you have shared from your course may also access these `clientFilesQuestion` files. This means that if you have any questions from your course that are publicly shared, anyone with access to the internet can access the `clientFilesQuestion` directories for these questions. Additionally, anyone with access to a question shared from your course will be able to access the client assets in your custom course elements, just as any students with access to your course can access the client assets in your custom course elements.
```

File: apps/prairielearn/src/pages/elementFiles/elementFiles.js (hunk 2 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js]: Discussed in a meeting, we aren't going to add these extra checks, because we aren't doing any of that for elementFiles within the same course anyway.

File: apps/prairielearn/src/question-servers/freeform.js (hunk 3 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/question-servers/freeform.js]: done, thanks

**Comment by SethPoulsen** [comment]
I'm getting a circular dependency error now: 

```
error no-circular: apps/prairielearn/src/lib/assets.ts → 
      apps/prairielearn/src/pages/elementFiles/elementFiles.js →
      apps/prairielearn/src/lib/question-variant.ts →
      apps/prairielearn/src/question-servers/index.ts →
      apps/prairielearn/src/question-servers/freeform.js →
      apps/prairielearn/src/lib/assets.ts
```

This was brought about by the fact that I am now using `getQuestionCourse` in `elementFiles.js`, which has to be imported from `question-variant.ts`.

Anyone have recommendations on fixing?

I was surprised that `assets.ts` is importing `elementFiles.js` this appears to be so that it can add or remove asset cacheing based on being in dev mode or production mode.

File: apps/prairielearn/src/question-servers/freeform.js [hunk 5]
```diff
@@ -1566,8 +1573,11 @@ export async function render(
           ..._.mapValues(dynamicDependencies.coreElementScripts, (file) =>
             assets.coreElementAssetPath(file),
           ),
-          ..._.mapValues(dynamicDependencies.courseElementScripts, (file) =>
-            assets.courseElementAssetPath(course.commit_hash, locals.urlPrefix, file),
+          ..._.mapValues(
+            dynamicDependencies.courseElementScripts,
+            (file) =>
+              assets.courseElementAssetPath(course.commit_hash, locals.urlPrefix, file) +
+              (!idsEqual(question.course_id, course.id) ? `?variant_id=${variant.id}` : ''),
```
Comment (mwest1066) [inline apps/prairielearn/src/question-servers/freeform.js:1580]: Can we add some tests for this new behavior? Both checking whether we can access what we should, as well as can't access what we shouldn't.

**Comment by jonatanschroeder** [comment]
> Anyone have recommendations on fixing?

What I've done in the past to fix these was to split a library in two.

> I was surprised that `assets.ts` is importing `elementFiles.js` this appears to be so that it can add or remove asset cacheing based on being in dev mode or production mode.

I suggest moving the `applyMiddleware` function in `assets.ts` to a separate file, possibly in `middlewares`.

File: apps/prairielearn/src/question-servers/freeform.js (hunk 5 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/question-servers/freeform.js:1580]: ya...there's currently no tests on these URLs at all, as far as I can tell, but I can add a few simple ones.

**Comment by SethPoulsen** [comment]
> I suggest moving the applyMiddleware function in assets.ts to a separate file, possibly in middlewares 

Done. It seems to have worked pretty nicely.

File: apps/prairielearn/src/tests/sharedQuestionPreview.test.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    testFileDownloads(previewPageInfo, downloadFile, false);
+
+    // TODO: implement
+    // testElementClientFiles(previewPageInfo, customElement);
+  });
+
+  describe('Test Shared Question Previews Within a Course', async function () {
+    const consumingCourseData = syncUtil.getCourseData();
+    consumingCourseData.course.name = 'CONSUMING 101';
+    const consumingCourseResults = await syncUtil.writeAndSyncCourseData(consumingCourseData);
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:88]: for some reason, calling `writeAndSyncCourseData` here is giving the following error: 

```
/PrairieLearn/packages/postgres/src/pool.ts:236
      throw new Error('Connection pool is not open');
            ^


Error: Connection pool is not open
    at PostgresPool.getClientAsync (/PrairieLearn/packages/postgres/src/pool.ts:236:13)
    at PostgresPool.queryAsync (/PrairieLearn/packages/postgres/src/pool.ts:512:31)
    at PostgresPool.queryOneRowAsync (/PrairieLearn/packages/postgres/src/pool.ts:535:31)
    at PostgresPool.queryRow (/PrairieLearn/packages/postgres/src/pool.ts:901:32)
    at selectOrInsertCourseByPath (/PrairieLearn/apps/prairielearn/src/models/course.ts:130:16)
    at Object.syncOrCreateDiskToSql (/PrairieLearn/apps/prairielearn/src/sync/syncFromDisk.ts:131:24)
    at syncCourseData (/PrairieLearn/apps/prairielearn/src/tests/sync/util.ts:448:29)
    at Object.writeAndSyncCourseData (/PrairieLearn/apps/prairielearn/src/tests/sync/util.ts:475:29)
    at Suite.<anonymous> (/PrairieLearn/apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:88:36)
```
anyone know why?

File: apps/prairielearn/src/tests/instructorQuestions.test.ts [hunk 7]
```diff
@@ -196,5 +162,6 @@ describe('Instructor questions', function () {
 
     testQuestionPreviews(previewPageInfo, addNumbers, addVectors);
     testFileDownloads(previewPageInfo, downloadFile, true);
+    testElementClientFiles(previewPageInfo, customElement);
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/instructorQuestions.test.ts:165]: This is adding tests for custom element client files in standard question previews, which tests did not exist at all before.

File: apps/prairielearn/src/tests/sharedQuestionPreview.test.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      siteUrl,
+      baseUrl,
+      questionBaseUrl: baseUrl + `/course/${consumingCourse.id}/question`,
+      questionPreviewTabUrl: '/preview',
+      isStudentPage: false,
+    };
+
+    testQuestionPreviews(previewPageInfo, addNumbers, addVectors);
+
+    testFileDownloads(previewPageInfo, downloadFile, false);
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:100]: These are doing standard tests on shared questions being used under course URLs (which we didn't have before, but should have).

File: apps/prairielearn/src/tests/sharedQuestionPreview.test.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+    testElementClientFiles(previewPageInfo, customElement);
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:102]: This line is where we actually test the new functionality that this PR is for.

File: apps/prairielearn/src/tests/sharedQuestionPreview.test.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+      questionPreviewTabUrl: '/preview',
+      isStudentPage: false,
+    };
+
+    testQuestionPreviews(previewPageInfo, addNumbers, addVectors);
+
+    testFileDownloads(previewPageInfo, downloadFile, false);
+
+    // TODO: implement
+    // testElementClientFiles(previewPageInfo, customElement);
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:82]: uncommenting this will test the functionality that is being implemented in #9807

File: apps/prairielearn/src/tests/sharedQuestionPreview.test.ts
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:1]: This file is actually a rename of `publicQuestionPreiew.test.ts` but git didn't catch it because I also added code.

File: apps/prairielearn/src/tests/instructorQuestions.test.ts [hunk 11]
```diff
@@ -59,30 +76,14 @@ describe('Instructor questions', function () {
...
-    });
-    it('should contain the differentiatePolynomial question', function () {
-      const question = _.find(questions, {
-        directory: differentiatePolynomial.qid,
+
+    for (const testQuestion of testQuestions) {
+      it(`should contain the ${testQuestion.qid} question`, function () {
+        const foundQuestion = _.find(questions, { directory: testQuestion.qid });
+        assert.isDefined(foundQuestion);
+        testQuestion.id = foundQuestion.id;
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/instructorQuestions.test.ts:84]: somewhat unrelated refactoring (though it is sort of related--if I hadn't added the for loop to de-dupe the code, I would have had to copy paste that same block yet another time for the `customElement` question I added tests for).

File: apps/prairielearn/src/question-servers/freeform.js (hunk 5 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/question-servers/freeform.js:1580]: I added tests for existing functionality that should have had tests as well as for new functionality.

File: apps/prairielearn/src/tests/sharedQuestionPreview.test.ts (hunk 6 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:88]: ok this is very weird, its only happening locally? the CI seems to have passed fine

Comment (nwalters512) [reply apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:88]: I have no explanation for why this would work in CI. As for why this fails, you might have a bad mental model of how Mocha executes tests. It will first execute all `describe` blocks to discover tests, before/after hooks, etc, and only _then_ will it actually execute the hooks/tests as needed. So, the reason this fails is because your code here runs before the callback in `before('set up testing server', helperServer.before());` is ever executed. As a rule, `describe` callbacks should not have side effects and should not be `async` (as Mocha won't await them).

To fix this, you can move all of this to a `before` hook here, which will execute after the `before` hook on the parent `describe` suite.

Comment (nwalters512) [reply apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:88]: A thought about why this works in CI: are you running this locally with something like `yarn mocha src/tests/sharedQuestionPreview.test.ts`? This actually uses a subtly different execution model (it doesn't run multiple processes for parallel execution), which I think explains the difference. You can reproduce passing tests locally with `make test-prairielearn`, which is what is run in CI. In fact, if you grep through the output, you'll see that `Test Shared Question Previews Within a Course` doesn't appear in the output, meaning that these tests were never even run.

Comment (SethPoulsen) [reply apps/prairielearn/src/tests/sharedQuestionPreview.test.ts]: Thanks Nathan, you were right about everything :) 

I believe its all fixed now.

**Review by nwalters512 — COMMENTED** [review]
Not a complete review; sharing an initial thought before I go into meetings.

File: apps/prairielearn/src/pages/elementFiles/elementFiles.js
[the remark this answers is not in the document]
Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:1]: I guess checking for any shared question in a course would be a cheap query, checking for any question shared to a particular other course would require a join. I guess we need to figure out what the security model is.

[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:1]: > Do query params affect cacheing? I had originally put the variant_id in the middle of the URL, @jonatanschroeder told me to make it a query param instead so that it _would_ cache properly.

To clarify my note: I asked for it to be a query param *that would only be used in shared questions*, so that in non-shared questions it would cache properly.

> * Force courses to provide the names of elements that should be unconditionally shareable without auth.

I'd advocate for this option, maybe with an option in the `info.json` file for the element, though I'm also fine with enabling sharing if the course has any sharing in place.

> Checking for any shared question in a course might be an expensive DB query to do on every question access (or maybe not? we need to talk about it).

If you want a cheaper query, you can check if the course has a sharing name, which IMHO is probably a good heuristic.

[the remark this answers is not in the document]
Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:1]: Thanks for the clarification Jonaton, sorry about the misunderstanding.

[the remark this answers is not in the document]
Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:1]: in a meeting with @mwest1066  and @nwalters512 , we decided on this one: 

> Maybe do the same, but specifically check for a sharing set that would grant the consuming course access to the sharing course's elements, and couple this with a change to serve these element files from `/pl/course/2/sharedElements/course/1/...`

[the remark this answers is not in the document]
Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:1]: Done.

File: apps/prairielearn/src/pages/elementFiles/elementFiles.sql [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+      q.shared_publicly
+      AND course_id = $producing_course_id
+    UNION
+    SELECT
+      1
+    FROM
+      sharing_sets AS ss
+      JOIN sharing_set_courses AS ssc ON ss.id = ssc.sharing_set_id
+      JOIN sharing_set_questions AS ssq ON ss.id = ssq.sharing_set_id
+      JOIN questions AS q on q.id = ssq.question_id
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/elementFiles/elementFiles.sql:18]: Do we want to actually check if there is a question in the sharing set here? or get rid of these two lines and just check if the sharing set exists?

File: apps/prairielearn/src/server.js [hunk 13]
```diff
@@ -847,6 +860,18 @@ export async function initExpress() {
...
+    (await import('./pages/elementFiles/elementFiles.js')).default,
+  );
+  app.use(
+    '/pl/course_instance/:course_instance_id(\\d+)/sharedElements/course/:producing_course_id(\\d+)/instructor/elements',
+    (await import('./pages/elementFiles/elementFiles.js')).default,
+  );
+  app.use(
+    '/pl/course/:course_id(\\d+)/sharedElements/course/:producing_course_id(\\d+)/elements',
+    (await import('./pages/elementFiles/elementFiles.js')).default,
+  );
```
Comment (nwalters512) [inline apps/prairielearn/src/server.js:874]: Review the comments above this block; these routes exist solely for backwards-compatibility. We don't need to add them for brand-new routes.

File: apps/prairielearn/src/tests/accessibility/index.test.ts [hunk 14]
```diff
@@ -151,12 +155,14 @@ const SKIP_ROUTES = [
...
   '/pl/course/:course_id/cacheableElementExtensions/:cachebuster/*',
   '/pl/course/:course_id/cacheableElements/:cachebuster/*',
+  '/pl/course/:course_id/sharedElements/course/:producing_course_id/cacheableElements/:cachebuster/*',
   '/pl/course/:course_id/clientFilesCourse/*',
   '/pl/course/:course_id/course_admin/file_download/*',
   '/pl/course/:course_id/news_item/:news_item_id/*',
   '/pl/course/:course_id/question/:question_id/clientFilesCourse/*',
   '/pl/course/:course_id/question/:question_id/clientFilesQuestion/*',
   '/pl/course/:course_id/elements/*',
+  '/pl/course/:course_id/sharedElements/course/:producing_course_id/elements/*',
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/accessibility/index.test.ts:165]: I was keeping this list alphabetized; let's do the same with these changes.

File: apps/prairielearn/src/tests/helperQuestionPreview.ts [hunk 15]
```diff
@@ -313,3 +313,42 @@ export function testFileDownloads(
...
+        if (error) {
+          return callback(error);
+        }
+        if (response.statusCode !== 200) {
+          return callback(new Error('bad status: ' + response.statusCode));
+        }
+        page = body;
+        callback(null);
+      });
+    });
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/helperQuestionPreview.ts:349]: Let's use `node-fetch` instead of introducing new callback-based code. You can also collapse these all into a single test with multiple assertions so you don't need the `let ...` variables outside of the scope.

File: apps/prairielearn/src/tests/helperQuestionPreview.ts [hunk 16]
```diff
@@ -313,3 +313,42 @@ export function testFileDownloads(
...
+    it('should have downloaded a file with the contents of course-element.js', function () {
+      assert(page.includes('This text was added by a script.'));
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/helperQuestionPreview.ts:351]: Calling this "page" is kinda weird.

File: apps/prairielearn/src/tests/sharedQuestionPreview.test.ts [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    }
+  });
+
+  before('set up another course to consume shared questions from ', async () => {
+    const consumingCourseData = syncUtil.getCourseData();
+    consumingCourseData.course.name = 'CONSUMING 101';
+    await syncUtil.writeAndSyncCourseData(consumingCourseData);
+  });
+
+  describe('Test Public Question Previews', function () {
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:73]: Nit, this would better match our conventions (same for other tests below):

```suggestion
  describe('Public question previews', function () {
```

File: apps/prairielearn/src/pages/elementFiles/elementFiles.sql (hunk 12 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.sql:18]: It should be cheap enough to check for the existence of an actual shared question; I'm fine to leave this as is.

**Review by nwalters512 — COMMENTED** [review]
Make sure you reformat these files to reorder imports once you resolve the merge conflicts!

File: apps/prairielearn/src/server.js (hunk 13 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/server.js:874]: Is the comment actually correct? 

I thought these routes are also always being used in local dev as well.

Comment (nwalters512) [reply apps/prairielearn/src/server.js:874]: 🙈 I independently discovered the same thing while working on #9864. Yes, let's change the comment to more accurately reflect why these are here!

Comment (SethPoulsen) [reply apps/prairielearn/src/server.js:874]: Would you like to do the comment in that PR? That would seem to make more sense than doing it here, since that PR is specifically about fixing the situation.

File: apps/prairielearn/src/tests/sharedQuestionPreview.test.ts (hunk 17 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:76]: done.

File: apps/prairielearn/src/tests/helperQuestionPreview.ts (hunk 16 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/tests/helperQuestionPreview.ts:353]: fixed

File: apps/prairielearn/src/tests/accessibility/index.test.ts (hunk 14 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/tests/accessibility/index.test.ts:161]: It doesn't seem particularly alphabetized--even in the snippet here, you can see the `/elements` URL appears after the `/question` one, which is not in order. the `grading_job` one also seems out of order. 

I mostly noticed this because I tried to solve the problem just by running the list through the command line `sort` tool, but that made way more changes than intended Maybe we should have a separate PR to just do that to the whole list? 

Anway, I changed the newly added lines to be as in-order as I could given the situation.

File: apps/prairielearn/src/tests/sharedQuestionPreview.test.ts [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+      isStudentPage: false,
+    };
+
+    testQuestionPreviews(previewPageInfo, addNumbers, addVectors);
+
+    testFileDownloads(previewPageInfo, downloadFile, false);
+
+    testElementClientFiles(previewPageInfo, customElement);
+  });
+});
```
Comment (SethPoulsen) [inline apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:124]: In addition to resolving all the PR comments, the latest commits also added these few lines to test question previews under `course_instance` urls, another thing we haven't had any tests for before now

File: apps/prairielearn/src/tests/helperQuestionPreview.ts (hunk 15 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/tests/helperQuestionPreview.ts]: done

File: apps/prairielearn/src/tests/helperQuestionPreview.ts [hunk 19]
```diff
@@ -315,3 +316,34 @@ export function testFileDownloads(
...
+      assert.lengthOf(elemList, 1);
+    });
+
+    it('should download a file with the contents of course-element.js', async function () {
+      const fileUrl = locals.siteUrl + elemList[0].attribs.src;
+      const response = await fetch(fileUrl);
+      assert.equal(response.status, 200);
+      const fileContents = await response.text();
+      assert(fileContents.includes('This text was added by a script.'));
+    });
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/helperQuestionPreview.ts:347]: Nit: I view the declaration of variables outside of tests to be a bit of a smell. In general I think we should prefer "fatter" tests that might make more assertions.

```suggestion
    it('should download a file with the contents of course-element.js', async function () {
      const elemList = locals.$('script[src*="course-element.js"]');
      assert.lengthOf(elemList, 1);
      
      const fileUrl = locals.siteUrl + elemList[0].attribs.src;
      const response = await fetch(fileUrl);
      assert.equal(response.status, 200);
      const fileContents = await response.text();
      assert(fileContents.includes('This text was added by a script.'));
    });
```

File: apps/prairielearn/src/tests/instructorQuestions.test.ts [hunk 20]
```diff
@@ -109,32 +110,15 @@ describe('Instructor questions', function () {
...
-        idsEqual(question.id, differentiatePolynomial.id),
-      );
-      assert.lengthOf(elemList, 1);
-      assert.equal(differentiatePolynomial.qid, elemList[0].qid);
-      assert.equal(differentiatePolynomial.title, elemList[0].title);
-    });
+
+    for (const testQuestion of testQuestions) {
+      it(`should include ${testQuestion.qid} question`, function () {
+        elemList = questionData.filter((question) => idsEqual(question.id, testQuestion.id));
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/instructorQuestions.test.ts:116]: Does `elemList` need to be declared in a higher scope? AFAICT there are two tests in which it's used and they don't (and in fact shouldn't) share state with each other.

File: apps/prairielearn/src/server.js [hunk 21]
```diff
@@ -804,6 +804,19 @@ export async function initExpress() {
...
+  app.use(
+    '/pl/course_instance/:course_instance_id(\\d+)/instructor/sharedElements/course/:producing_course_id(\\d+)/cacheableElements/:cachebuster',
+    (await import('./pages/elementFiles/elementFiles.js')).default,
+  );
+  app.use(
+    '/pl/course/:course_id(\\d+)/sharedElements/course/:producing_course_id(\\d+)/cacheableElements/:cachebuster',
+    (await import('./pages/elementFiles/elementFiles.js')).default,
+  );
```
Comment (nwalters512) [inline apps/prairielearn/src/server.js:818]: Nit: let's move this after the `// Serve element statics. ...` comment?

**Review by nwalters512 — APPROVED** [review]
No blockers, but a number of nits that would be nice to address.

File: apps/prairielearn/src/pages/elementFiles/elementFiles.js [hunk 22]
```diff
@@ -55,8 +59,25 @@ router.get(
     let elementFilesDir;
     if (res.locals.course) {
       // Files should be served from the course directory
```
[the remark this answers is not in the document]
Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.js:61]: I actually got rid of this comment in the follow on PR (https://github.com/PrairieLearn/PrairieLearn/pull/9807)

File: apps/prairielearn/src/tests/instructorQuestions.test.ts (hunk 20 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/tests/instructorQuestions.test.ts:116]: fixed

File: apps/prairielearn/src/tests/helperQuestionPreview.ts (hunk 19 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/tests/helperQuestionPreview.ts:347]: done

File: apps/prairielearn/src/question-servers/freeform.js [hunk 23]
```diff
@@ -1540,10 +1541,28 @@ export async function render(
...
       );
       dependencies.courseElementStyles.forEach((file) =>
-        styleUrls.push(assets.courseElementAssetPath(course.commit_hash, locals.urlPrefix, file)),
+        styleUrls.push(
+          assets.courseElementAssetPath(
+            course.commit_hash,
+            locals.urlPrefix +
+              (!idsEqual(question.course_id, variant.course_id)
+                ? `/sharedElements/course/${course.id}`
+                : ''),
```
[the remark this answers is not in the document]
Comment (SethPoulsen) [reply apps/prairielearn/src/question-servers/freeform.js:1550]: done

File: apps/prairielearn/src/server.js (hunk 21 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/server.js:818]: done

File: apps/prairielearn/src/pages/elementFiles/elementFiles.sql
Comment (nwalters512) [inline apps/prairielearn/src/pages/elementFiles/elementFiles.sql:1]: You'll need to take care to exclude deleted questions in this query.

Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.sql:1]: what's the standard check for that? `WHERE q.deleted_at <> NULL`?

Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.sql:1]: oh I guess `deleted_at IS NULL` is probably more proper

Comment (SethPoulsen) [reply apps/prairielearn/src/pages/elementFiles/elementFiles.sql:1]: done
